middleware/http/wasm/basic/basic.go

```diff
@@ -137,3 +121,90 @@ func (m *Middleware) getNativeMetadata(metadata middleware.Metadata) (*middlewar
 
 	return &data, nil
 }
+
+type wazeroRequestHandler struct {
+	r        wazero.Runtime
+	compiled wazero.CompiledModule
+}
+
+// preflightCheckWasm check that the wasm contains functions exported by TinyGo.
+// TODO: Validate these without instantiation after tetratelabs/wazero#2788
+func (rh *wazeroRequestHandler) preflightCheckWasm(ctx context.Context) error {
+	mod, err := rh.r.InstantiateModule(ctx, rh.compiled, wazero.NewModuleConfig())
+	if err != nil {
+		return err
+	}
+	defer mod.Close(ctx)
+
+	for _, fn := range []string{wasmFuncMalloc, wasmFuncFree, wasmFuncRewrite} {
+		if mod.ExportedFunction(fn) == nil {
+			return errors.Wrapf(err, "missing exported function %s", fn)
+		}
+	}
+
+	return nil
+}
+
+func (rh *wazeroRequestHandler) requestHandler(h fasthttp.RequestHandler) fasthttp.RequestHandler {
```

Comment:
trivia

one thing interesting for future work is basically I know of no allocator in wasm that's thread/goroutine safe as they don't use atomics to update the heap. This means that serious performance would rely on a ready pull who have to do one request and either reset or be thrown away. instantiation cost as it is now is ok as the wasm is simple, but even then you should be careful about the memory settings as certain toolchains/compilers reserve a lot of memory by default.